### PR TITLE
[Concurrency] Introduce runtime detection of data races.

### PR DESCRIPTION
**Pull request by DougGregor**
Through various means, it is possible for a synchronous actor-isolated
function to escape to another concurrency domain and be called from
outside the actor. The problem existed previously, but has become far
easier to trigger now that `@escaping` closures and local functions
can be actor-isolated.

Introduce runtime detection of such data races, where a synchronous
actor-isolated function ends up being called from the wrong executor.
Do this by emitting an executor check in actor-isolated synchronous
functions, where we query the executor in thread-local storage and
ensure that it is what we expect. If it isn't, the runtime complains.
The runtime's complaints can be controlled with the environment
variable `SWIFT_UNEXPECTED_EXECUTOR_LOG_LEVEL`:

  0 - disable checking
  1 - warn when a data race is detected
  2 - error and abort when a data race is detected

At an implementation level, this introduces a new concurrency runtime
entry point `_checkExpectedExecutor` that checks the given executor
(on which the function should always have been called) against the
executor on which is called (which is in thread-local storage). There
is a special carve-out here for `@MainActor` code, where we check
against the OS's notion of "main thread" as well, so that `@MainActor`
code can be called via (e.g.) the Dispatch library's
`DispatchQueue.main.async`.

The new SIL instruction `extract_executor` performs the lowering of an
actor down to its executor, which is implicit in the `hop_to_executor`
instruction. Extend the LowerHopToExecutor pass to perform said
lowering.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
The Windows failure is an unrelated flaky test:
```
Swift(windows-x86_64) :: Concurrency/Runtime/async_task_yield.swift
```

**Review by kavon — APPROVED** [review]
This is great! 👍🏼

**Comment by DougGregor** [comment]
@swift-ci please smoke test macOS

File: lib/SIL/Verifier/SILVerifier.cpp [hunk 1]
```diff
@@ -4789,7 +4789,14 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
                         "Builtin.Executor in lowered SIL");
     }
   }
-  
+
+  void checkExtractExecutorInst(ExtractExecutorInst *EEI) {
+    if (EEI->getModule().getStage() == SILStage::Lowered) {
+      require(false,
+              "extract_executor instruction should have been lowered away");
+    }
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply lib/SIL/Verifier/SILVerifier.cpp:4797]: (I'll pick this up in a subsequent cleanup PR)

**Comment by yyaroshevich** [comment]
Hello there! Could please someone clarify how to trigger runtime checks of correct executors? 
I've tried launching program with proper environment variables, but it does not seem to do anything since the compiled code does hot have calls to `_checkExpectedExecutor`. 
I'm using XCode 14.1 and Swift 5.7, maybe there is missing compiler switch to emit such check, I can't get it from this PR. Thanks a lot in advance!

**Comment by yyaroshevich** [comment]
Sorry for bothering, read code more carefully and found out that there is `-enable-actor-data-race-checks` compiler switch which does what I expected.

**Pull request merged**